### DISCUSSION TITLE
Poll: API Naming Color "Channel" vs. "Component"

### DISCUSSION THREAD

**Discussion by robloo** [AvaloniaUI/Avalonia#7852] [General]
Adding a comment here to bump up the discussion for more votes. I'm going to end the poll after this week.

---

I'm struggling to decide which name to use in color-related APIs: "Channel" or "Component". I thought I would pose this question to the community in the hopes of reaching consensus. **Please vote below!**

 > "ColorChannel" vs "ColorComponent" vs "Dimension". "Component" is more user-friendly and matches with iOS for example, however it is less technical than "Channel". ColorSpectrum in WinUI uses the term "Component" but the ColorPickerSlider uses "Channel". "Dimension" is used in a few places in WinUI but that is the worst name and won't be used. Overall, direction is moving towards "ColorChannel" as the appropriate name.
   https://github.com/AvaloniaUI/Avalonia/issues/4179#issuecomment-1054653895

This needs to be finalized before the 11.0 release and the API is locked. That gives some time to decide. That also means some PRs like `ColorSpectrum` might change API in the coming weeks.

Personally, I was gravitating towards "Channel". It sounds more technical and is shorter in length. **However, I have a growing gut feeling that may not be the correct choice. Some of this feeling comes from the HSV color model. In RGB "channel" and "component" can both be used fairly interchangeably. In HSV however, I'm not sure "channel" makes sense as it isn't an additive/subtractive color model.** [Channel (digital image)](https://en.wikipedia.org/wiki/Channel_(digital_image))

I started digging up examples:

Summary:

| Example | "Channel" | "Component" |
|-|-|-|
| Windows | ✅ | ✅ |
| iOS/macOS | 🔳 | ✅ |
| Android | 🔳 | ✅ |
| Wikipedia | 🔳 | ✅ |
| Gimp | ✅ | ✅ |
| Krita | ✅ | ✅ |
| Photoshop | ✅ | 🔳 |
| HTML | ✅ | ✅ |

 1. Windows
     - WinUI ColorPicker is inconsistent: There is the [ColorPickerHsvChannel enum](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.colorpickerhsvchannel?view=winrt-22000) and there is also the [ColorSpectrumComponents enum](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.colorspectrumcomponents?view=winui-3.0).
     - The WinRT/UWP-based [Windows.UI.Color struct](https://docs.microsoft.com/en-us/uwp/api/windows.ui.color?view=winrt-22000) uses the term "Channel" which is probably where most of my original thinking comes from.
     - The WPF-based [System.Windows.Media.Color struct](https://docs.microsoft.com/en-us/dotnet/api/system.windows.media.color?view=windowsdesktop-6.0) uses the term "Channel" -- this is the time frame Avalonia originated.
     - The oldest example from .NET [System.Drawing.Color struct](https://docs.microsoft.com/en-us/dotnet/api/system.drawing.color?view=net-6.0) uses the term "Component".
     - Did I mention how inconsistent Windows is? It seems originally the term "Component" was used back in the WinForms/GDI days. Then around WPF "Channel" became the term to use and that persisted until it was briefly forgotten for one enum on the UWP ColorPicker...

 2. iOS/macOS:
     * Apple has the UIColorPickerViewController. However, I couldn’t find examples in the API where this is applicable.
     * [UIColor](https://developer.apple.com/documentation/uikit/uicolor) uses the term “Component” in the docs.

 3. Android:
    * [android.graphics.Color](https://developer.android.com/reference/android/graphics/Color) Also uses the term “Components” in the docs.

 4. Wikipedia
    * Now uses the term “Components” pretty consistently https://en.wikipedia.org/wiki/Color_model
      > A color model is an abstract mathematical model describing the way [colors](https://en.wikipedia.org/wiki/Color) can be represented as tuples of numbers, typically as three or four values or color components. When this model is associated with a precise description of how the components are to be interpreted 
    * [Channel (digital image)](https://en.wikipedia.org/wiki/Channel_(digital_image)) May refer more to a greyscale image for a particular color channel.

 5. Photo Editors

    1. Gimp
        - In various places of the API such as [GimpRGB](https://developer.gimp.org/api/2.0/libgimpcolor/libgimpcolor-GimpRGB.html) Gimp seems to consistently use component.
        - Some filters like [Extract Component](https://docs.gimp.org/2.10/en/gimp-filter-component-extract.html) use "Component"
        - Other operations like [Channel Mixer](https://docs.gimp.org/2.10/en/gimp-filter-channel-mixer.html) use "Channel", interestingly it is within the "Components" menu.
	 
    2. Krita
        - In Python scripting API the term "Components" seems to be used https://scripting.krita.org/lessons/image-data
        - Within [Color Filters](https://docs.krita.org/en/reference_manual/filters/colors.html?highlight=channel#maximize-channel) the term "Channel" is used
        - The [Channels](https://docs.krita.org/en/reference_manual/dockers/channels_docker.html?highlight=channel#channels) docker uses the term "Channels" and is probably the most prominent use in the UI.
	 
    3. Photoshop
        - In the API's [SolidColor](https://developer.adobe.com/photoshop/uxp/2022/ps_reference/classes/solidcolor/), [RGBColor](https://developer.adobe.com/photoshop/uxp/2022/ps_reference/colors/rgbcolor/) it seems Photoshop cleverly avoids both terms and just uses the name + value. For example: "Red value".
        - In one place of the API "Channel" was found https://developer.adobe.com/photoshop/photoshop-api-docs/api/#operation/cutout

 6. HTML

     * Like Photoshop, avoids usage of the terms "Components" and "Channel" preferring instead to just say "value" in several places
     * The term "Alpha Channel" is used in some places of the [RGB color](https://www.w3schools.com/html/html_colors_rgb.asp) and [HSL and HSLA Colors](https://www.w3schools.com/html/html_colors_hsl.asp)
     * The term "Components" is used for [HEX Colors](https://www.w3schools.com/html/html_colors_hex.asp) docs

**Comment by robloo** [comment]
# Upvote here for the term "Channel"

 * Any down votes are ignored
 * Reply here to discuss "Channel" in more detail or provide your reasoning

**Comment by robloo** [comment]
# Upvote here for the term "Component"

 * Any down votes are ignored
 * Reply here to discuss "Component" in more detail or provide your reasoning

**Reply by moon6969** [reply]
My exposure to this is from using Gimp - my initial thoughts on seeing this poll is that a Gimp Channel is a container for a (color) Component (such as RGB), but it can also hold Alpha or Selection information.

Looking closer just now at the Gimp docs and UI blurs the line a little, but in general I think it uses Component.

**Reply by robloo** [reply]
Thanks for your feedback! Gimp is a great example and I'll look into it as well. Might look at Photoshop and Krita too along with their scripting.

**Reply by robloo** [reply]
@moon6969 I looked through Gimp (as well as a few others, see the updated description above). "Channel" was actually the most prominent term in the UI in the "Channel Mixer". "Component" was used in a few other operations and the API. It's hard to see anything but the "Channel Mixer" as the largest example although it is within the Components menu. I have to say Gimp is extremely inconsistent at least in the UI here.

 - In various places of the API such as [GimpRGB](https://developer.gimp.org/api/2.0/libgimpcolor/libgimpcolor-GimpRGB.html) Gimp seems to consistently use component.
 - Some filters like [Extract Component](https://docs.gimp.org/2.10/en/gimp-filter-component-extract.html) use "Component"
 - Other operations like [Channel Mixer](https://docs.gimp.org/2.10/en/gimp-filter-channel-mixer.html) use "Channel", interestingly it is within the "Components" menu.

**Comment by robloo** [comment]
Looks like "Component" wins 7-to-4. If anyone has strong objections to this for any reason: speak-up now! Otherwise, looks like the poll is closed.

Component is also more widely used in the review of examples.
